sdk/helper/storagepacker/ids2keys.go

```diff
@@ -0,0 +1,106 @@
+package storagepacker
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/vault/sdk/helper/cryptoutil"
+	"math"
+	"sort"
+	"strings"
+)
+
+type itemRequest struct {
+	// Item ID, provided by client
+	ID string
+
+	// Storage key == hash of ID
+	Key string
+
+	// Stored object, nil if not found
+	Value *Item
+}
+
+func GetItemIDHash(itemID string) string {
+	return hex.EncodeToString(cryptoutil.Blake2b256Hash(itemID))
+}
+
+// Given a list of IDs, calculate their keys generate itemRequests for each.
+func (s *StoragePackerV2) keysForIDs(ids []string) []*itemRequest {
+	requests := make([]*itemRequest, 0, len(ids))
+	for _, id := range ids {
+		requests = append(requests, &itemRequest{
+			ID:    id,
+			Key:   GetItemIDHash(id),
+			Value: nil,
+		})
+	}
+	return requests
+}
+
+// Given a list of Items, calculate their keys generate itemRequests for each.
+func (s *StoragePackerV2) keysForItems(items []*Item) []*itemRequest {
+	requests := make([]*itemRequest, 0, len(items))
+	for _, i := range items {
+		requests = append(requests, &itemRequest{
+			ID:    i.ID,
+			Key:   GetItemIDHash(i.ID),
+			Value: i,
+		})
+	}
+	return requests
+}
+
+// Sort the requests in key order, nondestructively (so we can refer
+// back to the original order.)
+func sortRequests(requests []*itemRequest) []*itemRequest {
+	duplicate := append([]*itemRequest{}, requests...)
+	sort.Slice(duplicate, func(i, j int) bool {
+		return duplicate[i].Key < duplicate[j].Key
+	})
+	return duplicate
+}
+
+func checkForDuplicateIds(ids []string) (bool, string) {
```

Comment:
Since this is only used by tests, should it go in a `_test.go` file?

Reply:
Also, I don't think it'll quite work as expected because, it's subtle, but if you instantiate a slice like this `make(map[string]bool, len(ids))` you have to insert things at specific indexes. Check this out: https://play.golang.org/p/bmVGCHkNo7Q

Reply:
Sorry, I don't get it; I thought make(map[string]bool, X) just preallocated space for X elements and was otherwise semantically equivalent to make(map[string]bool) or map[string]bool{}.

You're correct that this method didn't end up getting used in the implementation but maybe it should be:
* PutItem does its own check
* GetItem doesn't check at all and I'm not sure it works with duplicate IDs
* DeletedItem doesn't check but that's probably harmless

I should at least have test coverage for the first two.

Reply:
OK, there was coverage for PutItem already; I added tests for duplicated in GetItems and DeleteItems, and moved the test-only function to a test file.

Reply:
:+1: